File: 104_merge-k-sorted-lists/merge-k-sorted-lists.java
/*
@Copyright:LintCode
@Author:   heiyanbin
@Problem:  http://www.lintcode.com/problem/merge-k-sorted-lists
@Language: Java
@Datetime: 14-12-29 06:01
*/

/**
 * Definition for ListNode.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int val) {
 *         this.val = val;
 *         this.next = null;
 *     }
 * }
 */ 
public class Solution {
    /**
     * @param lists: a list of ListNode
     * @return: The head of one sorted list.
     */
    public ListNode mergeKLists(List<ListNode> lists) {  
        PriorityQueue<ListNode> pq = new PriorityQueue<ListNode>(10, 
        		new Comparator<ListNode>() {
        			public int compare(ListNode a, ListNode b) {
        				if (a == null) return -1;
        				if (b == null) return 1;
        				return a.val - b.val;
        			}
        		});
        for ( ListNode e : lists) {
        	if (e != null) pq.offer(e);
        }
        ListNode dummy = new ListNode(-1);
        for (ListNode prev = dummy; !pq.isEmpty(); ) {
        	prev.next = pq.poll();
        	prev = prev.next;
        	if (prev.next != null) pq.offer(prev.next);
        }
        return dummy.next;
    }
}

File: 104_merge-k-sorted-lists/merge-k-sorted-lists.cpp
/*
@Copyright:LintCode
@Author:   heiyanbin
@Problem:  http://www.lintcode.com/problem/merge-k-sorted-lists
@Language: C++
@Datetime: 15-05-29 16:34
*/

/**
 * Definition of ListNode
 * class ListNode {
 * public:
 *     int val;
 *     ListNode *next;
 *     ListNode(int val) {
 *         this->val = val;
 *         this->next = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param lists: a list of ListNode
     * @return: The head of one sorted list.
     */
    ListNode *mergeKLists(vector<ListNode *> &lists) {
        priority_queue<ListNode*, vector<ListNode*>, cmp> pq;
        for (auto &e : lists) {
            if (e) pq.push(e);
        }
        ListNode dummy(-1), *prev = &dummy;
        while (!pq.empty()) {
            prev->next = pq.top(); pq.pop();
            prev = prev->next;
            if (prev->next) pq.push(prev->next);
        }
        return dummy.next;
    }
    struct cmp {
        bool operator() (ListNode *a, ListNode *b) {
            return a->val > b->val;
        }
    };
};



